optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/ScoreDirectorFactoryFactory.java

```diff
@@ -95,18 +77,29 @@ public ScoreDirectorFactoryFactory(ScoreDirectorFactoryConfig config) {
 
     protected AbstractScoreDirectorFactory<Solution_, Score_> decideMultipleScoreDirectorFactories(
             ClassLoader classLoader, SolutionDescriptor<Solution_> solutionDescriptor) {
-        EasyScoreDirectorFactory<Solution_, Score_> easyScoreDirectorFactory =
-                buildEasyScoreDirectorFactory(solutionDescriptor);
-        AbstractConstraintStreamScoreDirectorFactory<Solution_, Score_> constraintStreamScoreDirectorFactory =
-                buildConstraintStreamScoreDirectorFactory(solutionDescriptor);
-        IncrementalScoreDirectorFactory<Solution_, Score_> incrementalScoreDirectorFactory =
-                buildIncrementalScoreDirectorFactory(solutionDescriptor);
-        DroolsScoreDirectorFactory<Solution_, Score_> droolsScoreDirectorFactory = buildDroolsScoreDirectorFactory(
-                classLoader, solutionDescriptor);
+        // Load all known Score Director Factories via SPI.
+        ServiceLoader<ScoreDirectorFactoryProvider> scoreDirectorFactoryProviders =
+                ServiceLoader.load(ScoreDirectorFactoryProvider.class);
+        Map<ScoreDirectorType, AbstractScoreDirectorFactory<Solution_, Score_>> factories =
+                new EnumMap<>(ScoreDirectorType.class);
+        for (ScoreDirectorFactoryProvider<Solution_, Score_> provider : scoreDirectorFactoryProviders) {
+            AbstractScoreDirectorFactory<Solution_, Score_> factory =
+                    provider.getScoreDirectorFactory(classLoader, solutionDescriptor, config);
+            if (factory != null) {
+                factories.put(provider.getSupportedScoreDirectorType(), factory);
+            }
+        }
+        AbstractScoreDirectorFactory<Solution_, Score_> easyScoreDirectorFactory = factories.get(EASY);
+        AbstractScoreDirectorFactory<Solution_, Score_> constraintStreamScoreDirectorFactory =
+                factories.get(CONSTRAINT_STREAMS);
+        AbstractScoreDirectorFactory<Solution_, Score_> incrementalScoreDirectorFactory = factories.get(INCREMENTAL);
+        AbstractScoreDirectorFactory<Solution_, Score_> droolsScoreDirectorFactory = factories.get(DROOLS);
 
+        // Make sure only one SDF is truly available.
```

Comment:
Let's not abbreviate to SDF - I'd argue that reads easily for the author for the next few months, but not for other developers.

Do we need a comment at all? I think the method name pretty much self documents